### PR TITLE
Harrison/new search

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

File: langchain/vectorstores/base.py [hunk 1]
```diff
@@ -324,13 +347,29 @@ def validate_search_type(cls, values: Dict) -> Dict:
...
         if "search_type" in values:
             search_type = values["search_type"]
-            if search_type not in ("similarity", "mmr"):
+            if search_type not in ("similarity", "similarity_score_threshold", "mmr"):
                 raise ValueError(f"search_type of {search_type} not allowed.")
+            if search_type == "similarity_score_threshold":
+                score_threshold = values["search_kwargs"].get("score_threshold")
+                if (score_threshold is None) or (
+                    not isinstance(score_threshold, float)
+                ):
```
Comment (pranjaldoshi96) [inline langchain/vectorstores/base.py:356]: Why separate condition for `None`, if isinstance for float is used then we should be able to validate it's None or not

```
>>> score_threshold  = None
>>> isinstance(score_threshold, float)
False
>>> score_threshold  = 0.15
>>> isinstance(score_threshold, float)
True
```

File: langchain/vectorstores/base.py [hunk 2]
```diff
@@ -124,10 +134,23 @@ def similarity_search_with_relevance_scores(
             similarity < 0.0 or similarity > 1.0
             for _, similarity in docs_and_similarities
         ):
-            raise ValueError(
+            warnings.warn(
```
Comment (eyurtsev) [inline langchain/vectorstores/base.py:137]: Do we have conventions about warnings vs. logger.wraning?

File: langchain/vectorstores/base.py [hunk 3]
```diff
@@ -116,6 +117,15 @@ def similarity_search_with_relevance_scores(
         """Return docs and relevance scores in the range [0, 1].
 
         0 is dissimilar, 1 is most similar.
+
+        Args:
+            query: input text
+            k: Number of Documents to return. Defaults to 4.
+            score_threshold: Optional, a floating point value between 0 to 1 to
```
Comment (eyurtsev) [inline langchain/vectorstores/base.py:124]: Could we expand kwargs to include this variable with a default of None?

Are we able to propagate this into the implementation itself? If there are use cases where `k` is large (say 10_000) it's nice to be able to have the data pre-filtered prior to being transmitted to the client.

Comment (eyurtsev) [inline langchain/vectorstores/base.py:124]: Would you mind clarifying in the doc-string that the filtering is happening client side?

**Review by eyurtsev — APPROVED** [review]
LGTM, but left comments with high effective temperature

**Comment by dev2049** [comment]
> LGTM, but left comments with high effective temperature

which is to say, given the emissivity of the comments, the real temperature is likely much higher

File: langchain/vectorstores/base.py (hunk 1 above)
Comment (jpzhangvincent) [reply langchain/vectorstores/base.py:356]: makes sense! will make the change

**Pull request merged**